include/drivers/dac.h

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2020 Libre Solar Technologies GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief DAC public API header file.
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_DAC_H_
+#define ZEPHYR_INCLUDE_DRIVERS_DAC_H_
+
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief DAC driver APIs
+ * @defgroup dac_interface DAC driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/** @brief DAC references. */
+enum dac_reference {
+	DAC_REF_DEFAULT,   /**< Default, reference cannot be selected. */
+	DAC_REF_VDD_1,     /**< VDD. */
+	DAC_REF_VDD_1_2,   /**< VDD/2. */
+	DAC_REF_VDD_1_3,   /**< VDD/3. */
+	DAC_REF_VDD_1_4,   /**< VDD/4. */
+	DAC_REF_INTERNAL,  /**< Internal. */
+	DAC_REF_EXTERNAL0, /**< External, input 0. */
+	DAC_REF_EXTERNAL1, /**< External, input 1. */
+};
+
+/**
+ * @brief Structure for specifying the configuration of a DAC channel.
+ */
+struct dac_channel_cfg {
+	/**
+	 * Channel identifier of the DAC that should be configured.
+	 */
+	u8_t channel_id;
+
+	/**
+	 * Desired resolution of the DAC (depends on device capabilities).
+	 */
+	u8_t resolution;
+
+	/**
+	 * Reference selection.
+	 */
+	enum dac_reference reference;
+
+	/**
+	 * Enable buffered output if available in underlying hardware.
+	 */
+	bool buffered;
```

Comment:
It’s a quite common feature. I would like to keep it in the the API.

Reply:
Could you give a few examples? Just to get a better feeling on how the feature is implemented / documented by different vendors.

Reply:
@mnkp Are you asking for examples of devices implementing a buffered DAC output? They are quite common. See the NXP Kinetis DAC32 for instance.

Reply:
Also Atmel SAM microcontrollers have a buffered output, see [here](http://ww1.microchip.com/downloads/en/AppNotes/atmel-42110-sam0-digital-to-analog-(dac)-driver_application%20note_at03244.pdf), Section 7.4.7. Configuration and Initialization (Channel). They specifically mention that the buffer should be switched off for power saving if not needed, so it has to be configurable at runtime.

Reply:
Thanks, yes I was asking for existing examples of how the "buffered output" feature is implemented by other vendors.

The NXP [DAC32: Digital-to-Analog Converter](https://mcuxpresso.nxp.com/api_doc/dev/1008/group__dac32.html) is a DAC API from NXP and when they talk about DAC buffer they mean a hardware buffer storing the samples. E.g. in case of K64 family it's a 16-word data buffer.

In the example given by @martinjaeger regarding SAM0 family DAC the buffered output refers to the possibility of enabling or disabling analog output buffer. E.g. with DAC buffer disabled the VOUT pin will not be driven, though the Vout voltage from the DAC can still be routed to an internal module.

As implemented in this PR the buffered option enables "higher drive current capability" of the analog output. If the option is disabled the analog output is still driven however with "lower drive current capability".

Every of the three examples talks about a different feature.

Reply:
@mnkp You are talking about the NXP DAC (not DAC32) present in the K64 family, which - as you say - does not implement a hardware buffer. I am talking about the NXP DAC32 present in e.g. the KE1xF series, which does implement a hardware buffer (and no, I can tell the difference between a hardware buffer and a data buffer).

Please see the following block diagram for the NXP DAC32 from the KE1xF reference manual which clearly shows an "AMP buffer":

<img width="547" alt="Screenshot 2020-03-30 at 07 10 58" src="https://user-images.githubusercontent.com/1076226/77877456-1f884680-7256-11ea-80c4-20a81e4caea9.png">

Reply:
On a side note, I have already implemented DAC drivers for the NXP DAC and DAC32 here:  https://github.com/henrikbrixandersen/zephyr/commits/mcux_dac_wip

Reply:
> Every of the three examples talks about a different feature.

I think it's the same thing for the mentioned Atmel SAM devices and the STM32. The SAM can only drive external pins with enabled buffer, but it's still a useful configuration option to switch it on or off. In case of STM32 you can drive an external pin even without buffer, but you get a much higher impedance output. So you can select if you need the output buffer depending on the application.

Reply:
Another option could be to implement the power management API for the DACs where the output can be turned off to save power - and convert the hardware buffer enable/disable to a devicetree setting.

Reply:
Thanks for the clarification. The behavior of NXP KE1xF looks similar to SAM0 family DAC in regards to a possibility of enabling/disabling the output buffer/amplifier. I checked a few more datasheets, it does look like a fairly popular feature.

Even if there are differences between vendors, like STM32 which allows to disable the buffer but still routes the analog signal to the output pad, it is a pretty consistent behavior.

However, the parameter description is a bit misleading at the moment
```
 * @param buffered Enable buffered output if available in underlying hardware.
```
We should be more explicit that we are talking about the analog output buffer. Also, every DAC will have some kind of analog buffer/amp that will drive the signal on the pad so user should always enable it if they intend to use it. E.g. in case of NXP K64 family it's not possible to disable the output buffer, as such the driver should return -ENOTSUP if the buffered parameter is set to false. If we believe that's not the intended behavior, i.e. if the driver doesn't support the feature it should be ignored this should also be documented.

For the usability it may be better to invert the logic and give user the option to disable the buffer rather than require him/her to explicitly enable it. e.g.
```
 * @param out_buf_disable Disable analog output buffer.
```